api/app/signals/apps/signals/managers.py

```diff
@@ -644,26 +649,77 @@ def update_type(self, data, signal):
 
         return signal_type
 
-    def _update_directing_departments_no_transaction(self, data, signal):
-        from signals.apps.signals.models.directing_departments import DirectingDepartments
+    def _update_user_signal_no_transaction(self, data, signal):
+        from signals.apps.users.models import User, SignalUser
+        try:
+            user_id = data['user_assignment']['user']['id']
+            signal.user_assignment = SignalUser.objects.create(
+                _signal=signal,
+                user=None if not user_id else User.objects.get(pk=user_id),
+                created_by=data['created_by'] if 'created_by' in data else None
+            )
+            signal.save()
+        except Exception:
+            pass
+        return signal.user_assignment
+
+    def _update_signal_departments_no_transaction(self, data, signal, relation_type):
+        from signals.apps.signals.models.signal_departments import SignalDepartments
+
+        relation = SignalDepartments.objects.create(
+            _signal=signal,
+            relation_type=relation_type,
+            created_by=data['created_by'] if 'created_by' in data else None
+        )
+
+        # check if different dep id is set, reset assigned user
+        if signal.user_assignment and relation_type == SignalDepartments.REL_ROUTING:
+            if signal.routing_assignment and signal.routing_assignment.departments.exclude(
+                id__in=[dept.id for dept in relation.departments.all()]
+            ).exists():
+                signal.user_assignment = None
 
-        directing_departments = DirectingDepartments.objects.create(_signal=signal, created_by=data['created_by'])
         for department_data in data['departments']:
-            directing_departments.departments.add(department_data['id'])
+            relation.departments.add(department_data['id'])
 
-        signal.directing_departments_assignment = directing_departments
+        if relation_type == SignalDepartments.REL_DIRECTING:
+            signal.directing_departments_assignment = relation
+        elif relation_type == SignalDepartments.REL_ROUTING:
+            signal.routing_assignment = relation
+        else:
+            raise Exception(f'Signal - department relation {relation_type} is not supported')
```

Comment:
`Exception` -> `ValidationError`  (The former will produce an HTTP 500 and the latter an HTTP 400 --- problem with the data should not cause an internal server error.)

Reply:
check